BUG: prevent ufunc output to an already-broadcasted array

mattip opened the pull request:
Fixes #2705. 

I tried to find a balance between fixing all possible edge cases and keeping the code simple. The documentation for [broadcast_arrays](http://www.numpy.org/devdocs/reference/generated/numpy.broadcast_arrays.html) already warns against writing to these, this PR attempts to detect a pending, perhaps unintentional write in a func operation and raises a `ValueError`

mattip committed:
BUG: prevent ufunc output to an already-broadcasted array

mattip commented:
It turns out not so trivial to find the equivalent of "writable operand with 0 stride and usable dim > 0."
I thought I had a good handle on this but

- Reductions and accumulations can set an output stride = 0
- Fancy indexing uses an iterator that never really iterates, it is used to calculate the strides.

seberg commented:
Some of that we can handle, but my guess is its either more complicated to detect (or needs a flag), or just doesn't belong in nditer but instead in the ufunc machinery in this case. Not sure though. Other pain points could be e.g. `ufunc.at` (but that is basically indexing).

mhvk commented:
This may indeed be better done in the ufunc setup - the fact that reduce uses this suggests that for the iterator one should continue to allow maximum freedom.

Also, a different and I think better solution is to just make broadcasted arrays read-only; this is indeed why `as_strided` has a `writable` flag added to it in numpy 1.12 and `broadcast_to` returns read-only arrays. The idea was to eventually do the same for `broadcast_arrays`, though it is not so easy to see how to do that...

mattip committed:
BUG: fix writing to broadcast at the ufunc level, but not all cases

mattip commented:
rebased to fix merge conflicts

mattip commented:
I moved the check to the ufunc level, added the trivial test from the issue and modified `memory_overlap` tests to verify that if the error is triggered then  `ufunc(b, c, out=out) != ufunc(b, c)`. This should cover most of the common cases, without trying too hard to systematically catch them all.

mhvk commented:
I'm still not sure about this - I think I'd prefer to just not generate any writable broadcast arrays by default, and leaving any other usage up to the user, i.e., rather than prevent doing things that are potentially risky and fragile, make it such that it is a conscious choice to do so.

mattip commented:
Closing. If any fix is needed it should be done in `broadcast_arrays`
